### ISSUE TITLE
Add horizontal skew and vertical skew features

### ISSUE THREAD

**Issue by guoyingtao** [guoyingtao/Mantis#44]
https://www.hackingwithswift.com/articles/135/how-to-render-uiviews-in-3d-using-catransformlayer
https://dzone.com/articles/a-look-at-perspective-transform-correction-with-co
https://stackoverflow.com/questions/8235288/perspective-correction-of-uiimage-from-points
https://github.com/paulz/PerspectiveTransform

`CATransform3D CATransform3DMakeRotation(CGFloat angle, CGFloat x, CGFloat y, CGFloat z);`

**Comment by guoyingtao** [comment]
Possible steps

1. Rotate UIImageView by CATransform3DMakeRotation
2. Get corner point of rotated UIImageView
3. Use CIPerspectiveTransform to create the new image

**Comment by guoyingtao** [comment]
```Swift
func CATransform3DMakePerspective(_ x: CGFloat, _ y: CGFloat) -> CATransform3D {
    var transform = CATransform3DIdentity
    transform.m34 = -1.0 / 1000.0
    transform = CATransform3DRotate(transform, y, 1, 0, 0)
    transform = CATransform3DRotate(transform, x, 0, 1, 0)
    return transform
}
```

**Comment by guoyingtao** [comment]
```Swift
@objc func handleSliderValueChanged(_ slider: UISlider) {
    let value = CGFloat(slider.value)
    let inputImage = CIImage(image: imageView.image!)
    
    // Calculate the perspective correction values
    let inputWidth = inputImage.extent.width
    let inputHeight = inputImage.extent.height
    let topLeft = CGPoint(x: 0, y: 0)
    let topRight = CGPoint(x: inputWidth, y: 0)
    let bottomLeft = CGPoint(x: 0, y: inputHeight)
    let bottomRight = CGPoint(x: inputWidth, y: inputHeight)
    let correction = CIPerspectiveTransform(inputImage: inputImage,
                                             topLeft: topLeft,
                                             topRight: CGPoint(x: topRight.x + value * inputWidth, y: topRight.y),
                                             bottomLeft: bottomLeft,
                                             bottomRight: CGPoint(x: bottomRight.x + value * inputWidth, y: bottomRight.y))
    
    // Apply the perspective correction to the CIImage
    let outputImage = correction.outputImage
    let context = CIContext()
    let cgImage = context.createCGImage(outputImage, from: outputImage.extent)
    imageView.image = UIImage(cgImage: cgImage!)
}
```

In this modified function, we calculate the perspective correction values based on the value of the slider, and then apply the CIPerspectiveTransform filter to the CIImage to create the corrected output image. Finally, we create a UIImage from the corrected output image and set it as the image for the UIImageView.

Note that this implementation assumes that the image is being cropped to preserve the original aspect ratio. If you want to allow for the image to be resized to fit the screen, you will need to adjust the calculation of the topRight and bottomRight points.

**Comment by guoyingtao** [comment]
Made a little bit progress with the code below

```Swift
        var transform = CATransform3DIdentity
        transform.m34 = -1.0 / 500.0
        transform = CATransform3DRotate(transform, totalRadians, 1, 0, 0)
        cropWorkbenchView.layer.transform = transform
```
![Simulator Screen Recording - iPhone 15 Pro Max - 2023-11-03 at 11 27 35](https://github.com/guoyingtao/Mantis/assets/26723384/610de879-4ab0-4537-86a9-75dc48b1b137)

**Comment by Karllas** [comment]
Hello,

Any plans of finishing this implementation?

**Comment by guoyingtao** [comment]
@Karllas 
I haven't found a good way to build this feature yet, so I have no plans to finish it in the near future.

**Comment by rickshane** [comment]
The solution to this problem is fairly straightforward in a Core Image workflow. There is a UX that presents two sliders to the user: a Horizontal Perspective adjustment slider with values ranging from -1 to +1, centered at 0, and a Vertical Perspective adjustment slider with the same values.

Both sliders will provide input to the Core Image CIPerspectiveTransform filter, which accepts as input four CIVectors that represent adjustments at the four corners of the image.

Only a single side of the image is adjusted for a given axis. A negative value for the horizontal adjustment results in the left side of the image being adjusted. A positive value for the horizontal adjustment results in the right side of the image being adjusted. As the adjustment in magnitude increase, both the x-axis and y-axis adjustment increases on the side of the adjustment (thought not necessarily with equal magnitude - this is something to be tuned by the implementer).
Likewise, for vertical adjustments, the same scenario applies. 

This is a stateless solution - subsequent adjustments do not stack as transactions, they merely replace the previous value.

A flip transform will invert any non-zero adjustment along the same axis, i.e., if there is a negative horizontal adjustment, then flipping the image horizontally will invert the horizontal adjustment to a positive value of the same magnitude.

A single function taking as input the two slider values then calculates the four input values to set in the CIPerspectiveTransform for upperLeft, upperRight, lowerLeft, lowerRight.

The most complex issue for Mantis is that of UX: how to design the user interface to accommodate the two sliders. Obviously the SlideControl may be used for both adjustments. They could be stacked in the area below the image where the rotation dial appears, or there could be a modal solution where only one control appears at a time.

**Comment by rickshane** [comment]
Here is the description of CITransformPerspective from Core Image Filter Reference:

![CITransformPerspective](https://github.com/guoyingtao/Mantis/assets/5288729/6ffdea9b-497c-45ea-8ca9-263656335b80)

**Comment by rickshane** [comment]
![Horizontal Perspective Adjustment](https://github.com/guoyingtao/Mantis/assets/5288729/b41c3c9a-90ff-4f87-9ec4-6c366227f69a)

**Comment by rickshane** [comment]
![Horizontal Perspective Adjustment](https://github.com/guoyingtao/Mantis/assets/5288729/08a4365b-562a-4bff-b3f4-071657fa4647)

**Comment by guoyingtao** [comment]
@rickshane 
Thanks for the suggestion!
The challenging part is when rotating the image, how to adjust cropBox for 3d space. I haven't figured it out yet.

**Comment by guoyingtao** [comment]
For UX part, we can directly use the design from Apple's Photos app which separated rotation with horizontal skew and vertical skew. 

I have another project https://github.com/guoyingtao/Inchworm which servers the similar purpose. Once I solved this issue, I can borrow the Apple's design to Mantis.

![IMG_8E9F12903487-1](https://github.com/guoyingtao/Mantis/assets/26723384/03d348bc-eaeb-4038-adde-0dc10d87e7aa)
![IMG_8AA9210E1D71-1](https://github.com/guoyingtao/Mantis/assets/26723384/ab9fa345-8ea2-4279-8c1b-7d62c666605e)

**Comment by rickshane** [comment]
I don't think the cropBox changes. After you adjust the perspective, you would crop the image to the original input extent. In your workflow, you would then create UIImage and add it back to the ImageContainer.

**Comment by rickshane** [comment]
Very nice component!

**Comment by rickshane** [comment]
I have been thinking about this problem again. 

It seems like the ideal solution would be to apply CATransform3DRotate to the CropWorkbenchView and not use the CIPerspectiveTransform filter. Otherwise, the image size would constantly be changing as the slider values change. You would then use CIPerspectiveTransform when you go to export the image since I don't think you can apply CATransform3DRotate to an image. But then given the same input for the x, y slider values, how would you match the output of the CropWorkbenchView transformation and the output of CIPerspectiveTransform filter when exporting the image? 

You would need the four corner values in the image to pass to CIPerspectiveTransform. Can you get this by a convert() call from CropAuxiliaryIndicatorView coordinates to the ImageContainerView coordinates?

**Comment by guoyingtao** [comment]
@rickshane 
I agree with you and that is also a possible solution what I think too.

There will be some math calculation work when doing image crop but the UI interaction part also need more work. I made a little bit progress to mimic the skew operation without any rotation (Which may need  the image keeps touching the cropBox corners while rotating in some scenarios)  

![Simulator Screen Recording - iPhone 15 Pro Max - 2024-05-12 at 00 55 22](https://github.com/guoyingtao/Mantis/assets/26723384/83cb15d1-835e-40b8-9742-69ddb4d44a99)

**Comment by rickshane** [comment]
Is this code in a branch or fork?

**Comment by guoyingtao** [comment]
No, it's just some test code and far from being usable

**Comment by rickshane** [comment]
When you rotate an image, are you changing the zoomScale and position of WorkbenchView to line up an edge? Where is that current logic in the code?

**Comment by rickshane** [comment]
I think this article might provide the solution to the skew step:

https://stackoverflow.com/questions/9470493/transforming-a-rectangle-image-into-a-quadrilateral-using-a-catransform3d/18606029#18606029

It provides a Swift class that can compute a CATransform3D that converts a rectangle to a quadrilateral based on the 4 corners of the quadrilateral. So this means that the two inputs from the sliders (horizontal and vertical) would be used to compute the four points of the transformed view (or the adjusted image). You could then pass those 4 inputs to the logic in the above article to transform the WorkbenchView and also use the same 4 inputs to pass to CIPerspectiveTransform during the crop step.

This solution is not using CATransform3DRotate.

**Comment by guoyingtao** [comment]
> When you rotate an image, are you changing the zoomScale and position of WorkbenchView to line up an edge (or edges) of a rectangle to the cropBox? Where is that current logic in the code?

Yes, but it is for the WorkbenchView without 3d transform, the logic is in the CropView.adjustWorkbenchView function.

```Swift
private func adjustWorkbenchView(by radians: CGFloat) {
        let width = abs(cos(radians)) * cropAuxiliaryIndicatorView.frame.width + abs(sin(radians)) * cropAuxiliaryIndicatorView.frame.height
        let height = abs(sin(radians)) * cropAuxiliaryIndicatorView.frame.width + abs(cos(radians)) * cropAuxiliaryIndicatorView.frame.height
        
        cropWorkbenchView.updateLayout(byNewSize: CGSize(width: width, height: height))
        
        if !isManuallyZoomed || cropWorkbenchView.shouldScale() {
            cropWorkbenchView.zoomScaleToBound(animated: false)
            isManuallyZoomed = false
        } else {
            cropWorkbenchView.updateMinZoomScale()
        }
        
        cropWorkbenchView.updateContentOffset()
    }
```

**Comment by rickshane** [comment]
I created a topic branch that contains some basic logic (UX and state management) to support three rotation types (straighten, vertical skew, horizontal skew) in the EmbeddedViewController. Selecting Straighten works as expected, Selecting horizontal or vertical skew does nothing (other than setting some state). Note: I do not maintain rotation degrees state for each type yet. 

Here is a video:


https://github.com/guoyingtao/Mantis/assets/5288729/91b11704-cd8b-4043-967c-d261b0d642e3

I will issue a pull request in case you want to use some of this code.

**Comment by rickshane** [comment]
> > When you rotate an image, are you changing the zoomScale and position of WorkbenchView to line up an edge (or edges) of a rectangle to the cropBox? Where is that current logic in the code?
> 
> Yes, but it is for the WorkbenchView without 3d transform, the logic is in the CropView.adjustWorkbenchView function.
> 
> ```swift
> private func adjustWorkbenchView(by radians: CGFloat) {
>         let width = abs(cos(radians)) * cropAuxiliaryIndicatorView.frame.width + abs(sin(radians)) * cropAuxiliaryIndicatorView.frame.height
>         let height = abs(sin(radians)) * cropAuxiliaryIndicatorView.frame.width + abs(cos(radians)) * cropAuxiliaryIndicatorView.frame.height
>         
>         cropWorkbenchView.updateLayout(byNewSize: CGSize(width: width, height: height))
>         
>         if !isManuallyZoomed || cropWorkbenchView.shouldScale() {
>             cropWorkbenchView.zoomScaleToBound(animated: false)
>             isManuallyZoomed = false
>         } else {
>             cropWorkbenchView.updateMinZoomScale()
>         }
>         
>         cropWorkbenchView.updateContentOffset()
>     }
> ```

Is this a correct statement to the problem?:  

You have currently solved how to align a rotated rectangle (the cropWorkBenchView.frame) along a bounding rectangle (the CropBox). Now you need to solve how to align a rotated equilateral trapezoid (which is just a rectangle with two equilateral triangles on each side, with known vertex lengths and angles) along a bounding rectangle (the CropBox). 

Is that the problem that needs to be solved?

**Comment by guoyingtao** [comment]
Looks like it is.
The 3d interaction demo I was showing includes only skew + special translation. When adding scale, rotation and random translation, we still need to keep the CropBox inside the image, and sometimes the images edges need to touch the CropBox vertices while rotating. 
I haven't figured it out a good solution and haven't got enough time to work on it.

**Comment by rickshane** [comment]
Does your 3d interaction demo use "CATransform3DRotate" ?

**Comment by guoyingtao** [comment]
It uses CATransform3DRotate + CATransform3DTranslate (When rotation angle is greater than a specified angle, it begins to move the rotation axis with CATransform3DTranslate to mimic the similar Photo.app's skew interactions on iPhone.)

**Comment by rickshane** [comment]
Does the translate start getting called at +/- 10 degrees?

**Comment by rickshane** [comment]
I updated the PR to add state management that updates RotationDial value when changing modes between straighten, horizontal skew, and vertical skew. 

I also added basic horizontal and vertical skewing by applying a transform to ImageContainer instead of CropWorkbenchView (applying transform to CropWorkbenchView gave unexpected results). For now, I only skew one side of image depending on whether slider value is positive or negative. One bug I cannot figure out is that the opposite side of the skew is also skewing. This is strange, because I have a separate test harness that does not use Mantis code and the non-skewed side stays at original anchor points.

This skew method does not use CATransform3D rotate or translate. I will try to add some code to skew opposite side when degrees >= threshold (+/- 10 degrees?) to look more like Apple Photos app.

I notice that the contentOffset is in the wrong place if I skew image and then try to straighten the image (rotate). So cannot switch between rotating and skewing at this point.

Here is a video:


https://github.com/guoyingtao/Mantis/assets/5288729/c5d12cfa-f005-406d-b1f8-c264eb18e1c4

**Comment by rickshane** [comment]
I just updated the PR to apply the CATransform3D to the CropWorkbenchView.layer instead of the ImageContainer.layer. I just needed to normalize the coordinates against the cropBoxFrame.

So CropWorkbenchView now transforms as expected. But the ImageContainer is not correct. Perhaps the transform must be applied to ImageContainer.layer.transform, as well. 

I made the CropWorkbenchView.background color = red to show the disparity between CropWorkbenchView and ImageContainer.

Here is a video:


https://github.com/guoyingtao/Mantis/assets/5288729/e471c3fe-b827-49e3-9067-166a49f6c530

**Comment by rickshane** [comment]
I wonder if the problem between skew and straighten is that straighten adjustments are CGAffineTransform applied to CropWorkbenchView.transform and skew adjustments are CATransform3D applied to CropWorkbenchView.layer.transform.

Perhaps both types of adjustments must be same type?

**Comment by rickshane** [comment]
I just updated the PR to apply the skew transform to the ImageContainer after applying it to the CropWorkbenchView.

It looks like this is the correct technique. It works perfectly in my other test harness. However in Mantis, it is not perfectly lined up. There is still some offset bug I need to figure out.

Here is a video:


https://github.com/guoyingtao/Mantis/assets/5288729/98413c99-ef9c-4770-9065-fc07b8cdd69a

**Comment by rickshane** [comment]
> Looks like it is. The 3d interaction demo I was showing includes only skew + special translation. When adding scale, rotation and random translation, we still need to keep the CropBox inside the image, and sometimes the images edges need to touch the CropBox vertices while rotating. I haven't figured it out a good solution and haven't got enough time to work on it.

Would you please provide a video example from Photos app where rotating a skewed image is touching the CropBox and CropBox stays inside the image?

Thanks

**Comment by guoyingtao** [comment]
@rickshane 
Sure. You can see the example below. Also thanks for working on this issue.

https://github.com/guoyingtao/Mantis/assets/26723384/cf6869db-dc74-44d0-943b-cf8ac1a6a8dc

**Comment by rickshane** [comment]
Fixed imageView alignment issue when skewing image. Commented out call to set imageView.contentMode to .scaleAspectFit seems to fix this bug.

Video:


https://github.com/guoyingtao/Mantis/assets/5288729/6a00cff8-e340-44fa-a59b-7c4f6d94731a

**Comment by rickshane** [comment]
@guoyingtao 

I just noticed that Mantis rotates in opposite direction to Apple Photos app. Is this by design?

**Comment by guoyingtao** [comment]
> @guoyingtao 
> 
> 
> 
> I just noticed that Mantis rotates image in opposite direction to Apple Photos app. Is this by design?
> 
> 
> 
> In fact, most editing apps rotate image in opposite direction to Apple Photos app.
> 
> 
> 
> 

I didn't notice it before, but I feel like the rotating direction is more natural in Mantis than Apple Photo app. 
When you move rotation dial, you will feel that the photo rotates to the same direction with your finger which I think is more natural.

**Comment by rickshane** [comment]
@guoyingtao 

It looks like this function in CropWorkbenchView is never called. May it be removed?

func shouldScale() -> Bool {
        return contentSize.width / bounds.width <= 1.0
            || contentSize.height / bounds.height <= 1.0
    }

**Comment by guoyingtao** [comment]
It is used in the function `adjustWorkbenchView` in `CropView`
<img width="1010" alt="image" src="https://github.com/guoyingtao/Mantis/assets/26723384/2a57ace6-e691-4f79-98db-4180bc26b43f">

**Comment by rickshane** [comment]
@guoyingtao

My mistake. Apologies.

I am still trying to understand the math for resizing the cropWorkbenchView upon rotation to align the auxiliaryIndicatorView. The code is almost similar to the formula for rotating an image using a matrix transform except one operator is positive instead of negative.

Please explain the first two lines in adjustWorkbenchView():

let width = abs(cos(radians)) * cropAuxiliaryIndicatorView.frame.width +
abs (sin(radians)) * cropAuxiliaryIndicatorView.frame.height
let height = abs(sin(radians)) * cropAuxiliaryIndicatorView.frame.width +
abs (cos (radians)) * cropAuxiliaryIndicatorView.frame.height

How does this work? Is this a standard formula? Is there a reference somewhere explaining it?

**Comment by guoyingtao** [comment]
@rickshane 
I did the math and no reference for it.
You can check the drawing below for the basic ideas.

<img width="1012" alt="image" src="https://github.com/guoyingtao/Mantis/assets/26723384/952ea382-6e2a-460d-893b-d09924547dd5">

**Comment by rickshane** [comment]
Thanks for the diagram. 

The math made sense once I hid the ImageContainer and saw the bounds change of the CropWorkbenchView.

**Comment by rickshane** [comment]
@guoyingtao

I just updated the PR with added support for concatenating rotation and skew operations. 

I have disabled the flip and adjustCropWorkbenchView calls until those are solved.

Here is a video update:

https://github.com/guoyingtao/Mantis/assets/5288729/446daf86-fbd6-42bc-84bb-2dfa38eb0c73

**Comment by Karllas** [comment]
@rickshane @guoyingtao is this PR planned to be finished anytime soon?

**Comment by guoyingtao** [comment]
@Karllas 
The UI interactions is complicated so we haven't figured it out a good solution yet. So this feature won't be finished anytime soon.
